Introducing: Java client v3

Ant-hem opened the pull request:
# Introducing Java Client v3 ☕️

# ChangeLog/ New features

## Maven

To get the beta version add this to your pom.xml

```xml
<dependency>
         <groupId>com.algolia</groupId>
        <artifactId>algoliasearch-core</artifactId>
        <version>3.0.0-beta-2</version>
 </dependency>

<dependency>
        <groupId>com.algolia</groupId>
         <artifactId>algoliasearch-apache</artifactId>
         <version>3.0.0-beta-2</version>
</dependency>
```

## Less dependencies!
- org.apache.httpcomponents.httpasyncclient
- com.fasterxml.jackson.core / jackson-core / jackson-databind /jackson-datatype-jsr310
- And that's it...

## Injectable HttpCient
The default HttpClient is httpasyncclient from Apache. If you want another one, just implement the `
IHttpRequester` interface with your favorite HttpClient and inject it!

## New test suite!
- Powered by JUnit5
- Running in less than 30 seconds locally! ⚡️

## Asynchronous and synchronous methods in the same class!

Before, asynchronous and synchronous methods were split in two different classes. Now, they are reunited in the same class! 

All asynchronous methods are suffixed with the `async` keyword and return a `CompletableFuture`.

```java
SearchClient searchClient = DefaultSearchClient.create(ALGOLIA_APPLICATION_ID_1, ALGOLIA_API_KEY_1);
SearchIndex<Employee> index = searchClient.initIndex("employees", Employee.class);
    
// Async Code
CompletableFuture<SearchResult<Employee>> searchFuture = index.searchAsync(new Query("algolia"));

// Sync code
SearchResult<Employee> search = index.search(new Query("algolia"))
```
## Useful helper: waiTask()

All Algolia's asynchronous operations can now be awaited like this:

```java
index.setSettings(indexSettings).waitTask();
```

## Static factories

We added static factories to Algolia's POJOs such as `Synonyms`, `Rules` to help you build valid Algolia's objects.

```java
    Synonym wiiToWiiu =
        Synonym.createOneWaySynonym("wii_to_wii_u", "wii", Collections.singletonList("wii u"));
```

# Getting started locally:
- Clone the repo and checkout on the `develop` branch.

- Build the project
```sh
mvn clean install
```

- Run the tests
```sh
mvn clean test
```

Ant-hem committed:
Scaffolding: v3 project

https://github.com/algolia/algoliasearch-client-java-2/projects/1#card-17598839
https://github.com/algolia/algoliasearch-client-java-2/projects/1#card-16731548
https://github.com/algolia/algoliasearch-client-java-2/projects/1#card-16731615

Ant-hem committed:
Added: HttpRequester and HttTransport layer.

https://github.com/algolia/algoliasearch-client-java-2/projects/1#card-17598839
https://github.com/algolia/algoliasearch-client-java-2/projects/1#card-16731548
https://github.com/algolia/algoliasearch-client-java-2/projects/1#card-16731615

Ant-hem committed:
Updated: The transport layer.

Added extraHeaders support
Added extraQueryParams support
Added a Better exception handling
Improved the recursive retry with thenCompose

https://github.com/algolia/algoliasearch-client-java-2/projects/1#card-17598839
https://github.com/algolia/algoliasearch-client-java-2/projects/1#card-16731548
https://github.com/algolia/algoliasearch-client-java-2/projects/1#card-16731615

Ant-hem committed:
Added: Launder throwable and Executor service

Added the possibility to inject an executor service
https://github.com/algolia/algoliasearch-client-java-2/projects/1#card-17979337

Improved exception handling with a launderThrowable to allow users to catch more easily business exceptions.
https://github.com/algolia/algoliasearch-client-java-2/projects/1#card-17979337

Ant-hem committed:
Added: SearchIndex class and WaitableResponse logic

Added the possibility to inject an executor service
https://github.com/algolia/algoliasearch-client-java-2/projects/1#card-16731666

Ant-hem committed:
Removed: IRetryStrategy and IHttpTransport

Ant-hem committed:
Removed: IRetryStrategy and IHttpTransport

Ant-hem committed:
Scaffolding: AnalyticsClient

Ant-hem committed:
Scaffolding: InsightsClient

Ant-hem committed:
Added: Analytics client

Ant-hem committed:
Added: InsightsClient async methods

Ant-hem committed:
Added: search methods and JavaType in transport layer

Improved the transport layer to handle generics return class

Ant-hem committed:
Removed: unnecessary requestOptions allocations

Ant-hem committed:
Added: multipleGetObjects

Ant-hem committed:
Code refactoring

- Added null check
- Added overloaded method in httpTransport
- Added more precise type in AlgoliaHttpRequest

Ant-hem committed:
Added: getSettings

Ant-hem committed:
Added: getRule

Ant-hem committed:
Added: searchRule

Ant-hem committed:
Added: deleteObject

Ant-hem committed:
Added: clearObjects

Ant-hem committed:
Added: multipleBatch

Ant-hem committed:
Added: multipleQueries

ElPicador commented on algoliasearch/pom.xml:
```diff
@@ -25,6 +25,11 @@
+    <dependency>
+      <groupId>org.asynchttpclient</groupId>
+      <artifactId>async-http-client</artifactId>
+      <version>2.7.0</version>
```
it's a good practice to put the version in a property, at the beginning of the file

ElPicador commented on algoliasearch/src/main/java/com/algolia/search/clients/AlgoliaConfig.java:
```diff
@@ -0,0 +1,121 @@
+    return applicationID;
+  }
+
+  public AlgoliaConfig setApplicationID(String applicationID) {
```
😢, mutability

Ant-hem replied:
Good catch! I shouldn't be too confident with intelij generators 😅

ElPicador commented on algoliasearch/src/main/java/com/algolia/search/clients/InsightsConfig.java:
```diff
@@ -0,0 +1,31 @@
+public class InsightsConfig extends AlgoliaConfig {
+
+  public InsightsConfig(String applicationID, String apiKey) {
+    this(applicationID, apiKey, "us");
```
Only `us`?

Ant-hem replied:
`US` is the default one at the moment isn't it? 
That's why we provide another constructor to override it, if it's not the good one.

ElPicador replied:
👌

ElPicador commented on algoliasearch/src/main/java/com/algolia/search/clients/AnalyticsClient.java:
```diff
@@ -0,0 +1,314 @@
+      long id, RequestOptions requestOptions) {
+    return transport.executeRequestAsync(
+        HttpMethod.POST,
+        "/2/abtests/" + id + "/stop",
```
I would make sure `id` is properly escaped with a URL encoding. Same comment for everytime we are building URLs

julienbourdeau replied:
I'd recommend something similar to what I did in PHP. I defined a helper method to escape all variables. It looks like a sprintf call.

```php
api_path("/2/abtests/%s/stop", id)
api_path('/1/indexes/%s/facets/%s/query', indexName, facetName),
```

https://github.com/algolia/algoliasearch-client-php/blob/master/src/Support/Helpers.php#L26-L35

Ant-hem replied:
Thanks 👍 I have did the same in .NET client. I will add it there.

ElPicador commented on algoliasearch/src/main/java/com/algolia/search/clients/InsightsConfig.java:
```diff
@@ -0,0 +1,31 @@
+public class InsightsConfig extends AlgoliaConfig {
+
+  public InsightsConfig(String applicationID, String apiKey) {
+    this(applicationID, apiKey, "us");
```
only `US`?

ElPicador commented on algoliasearch/src/main/java/com/algolia/search/helpers/QueryStringHelper.java:
```diff
@@ -0,0 +1,25 @@
+    }
+  }
+
+  public static String buildQueryString(Map<String, String> map) {
```
no escaping of values?

Ant-hem committed:
Updated: AlgoliaConfig class

Ant-hem committed:
Added: splitIntoBatches

Ant-hem committed:
Fixed: getAbtests wrong http verb and calltype

Ant-hem committed:
Fixed: synchronized statement in the retry strategy

Ant-hem committed:
Added: searchClient getLogsAsync

Ant-hem committed:
Scaffolding: Common test suite

Ant-hem committed:
Added: SearchIndex.search

ElPicador commented on algoliasearch/src/main/java/com/algolia/search/clients/AnalyticsClient.java:
```diff
@@ -0,0 +1,314 @@
+  public AnalyticsClient(@Nonnull AnalyticsConfig config, @Nonnull IHttpRequester httpRequester) {
+
+    Objects.requireNonNull(httpRequester, "An httpRequester is required.");
+    Objects.requireNonNull(config, "A configuration is required.");
```
No needed as the parameters are already marked as @Nonnull

Ant-hem replied:
`@NonNull` is only an annotation and won't throw an exception at runtime. I think it's better to have the double check to fail fast. IMO the combination of both is complementary.

ElPicador commented on algoliasearch/src/main/java/com/algolia/search/clients/AlgoliaConfig.java:
```diff
@@ -0,0 +1,110 @@
+    this.defaultHeaders =
+        new HashMap<String, String>() {
+          {
+            put("X-Algolia-Application-Id", getApplicationID());
```
I would use constants for constant strings

ElPicador commented on algoliasearch/src/main/java/com/algolia/search/clients/AnalyticsConfig.java:
```diff
@@ -0,0 +1,27 @@
+    List<StatefulHost> hosts =
+        Collections.singletonList(
+            new StatefulHost(
+                "analytics.algolia.com",
```
I would use the regional endpoint. analytics.us.algolia.com

As we have more than one

ElPicador reviewed (commented):
Really nice improvements 👍

Ant-hem committed:
Added: saveObjects

Ant-hem committed:
Added: partialUpdateObjectsAsync

Ant-hem committed:
Added: partialUpdateObjectAsync

Ant-hem committed:
Added: settings test

Ant-hem committed:
Added: Date serialization/deserialization support with jackson-datatype

Ant-hem committed:
Code cleaning

Ant-hem committed:
Added: Analytics ABTest test

- Also added global teardown to delete all the indices
- Added ABTest response class

Ant-hem committed:
Added: rules and synonyms methods

- saveRule,saveRules,replaceAllRules
- searchSynonym,getSynoym,saveSynonym(s),replaceAllSynonyms

redox commented on algoliasearch-common/src/main/java/com/algolia/search/Defaults.java:
```diff
@@ -19,7 +20,10 @@
+          .registerModule(new JavaTimeModule())
           .enable(OBJECT_MAPPER_DEFAULT_FEATURE)
+          .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
+          .enable(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY)
```
Could you add some comments/docs here, why enabling & disabling those?

Ant-hem replied:
Fixed with eeb6b15! TLDR;

`DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY`: Enabled to allow Jackson to deserialize List/Array of one element.

`SerializationFeature.WRITE_DATES_AS_TIMESTAMPS`: Disabled because most of the dates we **send** to the API are ISO String.

However, those can be overridden with annotations!

redox replied:
> `SerializationFeature.WRITE_DATES_AS_TIMESTAMPS`: Disabled because most of the dates we **send** to the API are ISO String.

Since JSON doesn't support date, we should always expect either a string (but we should not do any parsing whatsovever) or an integer I think.

redox replied:
So, OK for converting them to timestamp I think

Ant-hem committed:
Updated: serializer settings and custom serializers

- Added query string serializer
- Updated serializer documentation
- Updated assertj version

Ant-hem committed:
Added: multipleOperations tests

Ant-hem committed:
Added: APIKey methods

- deleteApiKeys
- getApiKey
- restoreApiKey
- updateApiKey
- addApiKey

Ant-hem committed:
Added: apiKeys test

Ant-hem committed:
Added/fixed: query rules tests

- Rules iterators
- Rules POJOS
- Rules custom serializer/deserializer
- Rules methods error fix

Ant-hem committed:
Updated: fmt-maven-plugin to 2.0 to 2.8

Ant-hem committed:
Added: Synonyms methods and iterators

- searchSynonyms
- clearSynonyms
- deleteSynonym
- synonyms iterator

Ant-hem committed:
Added: Synonyms tests

Ant-hem committed:
Code: formatting

Ant-hem committed:
Updated: coverage dependencies & script

Ant-hem committed:
Added: index iterator

Ant-hem committed:
Added: indexing test and methods

- added getObjects
- added deleteObjects
- added Algolia helper for reflection

Ant-hem committed:
Added: accountClient feature and related tests

Ant-hem committed:
Added: search test

- added searchForFacetValues

Ant-hem committed:
Added: replacing test and methods

- added replaceAllObjects
- added copyToAsync
- added moveFromAsync

Ant-hem committed:
Added: batching test

- added static factories for batch request

Ant-hem committed:
Added: synonyms static factories

Ant-hem committed:
Added: insights test

Ant-hem committed:
Added: connect timeout

Ant-hem committed:
Cleanup: exceptions

Ant-hem committed:
Added: MCM methods

- listClustersAsync
- searchUserIDsAsync
- listUserIDsAsync
- getUserIDAsync
- getTopUserIDAsync
- removeUserIDAsync
- assignUserIDAsync

Ant-hem committed:
Removed dependencies to V2

This commit makes the v3 independent from the v2
It removes also also the external dependencies

Ant-hem committed:
Code refactoring

Ant-hem committed:
Code cleanup

Ant-hem committed:
Code refactoring: Split SearchIndex class into multiple classes

Ant-hem committed:
Code refactoring: folder organizations for protection level

Ant-hem committed:
Added: MCM tests

Ant-hem committed:
Added: serializer for removeStopWords and typoTolerance

Ant-hem committed:
Added: generateSecuredAPIKey feature and test

Ant-hem committed:
Added: personalization features and test

Ant-hem committed:
Added: copyIndex feature and test

Ant-hem committed:
Added: facetFilters serializer test

Ant-hem committed:
Refactor: test suite

Ant-hem committed:
Changed: asyncHttpClient to Apache HttpAsyncClient

Ant-hem committed:
Updated: Exceptions documentation

Ant-hem committed:
Deleted: legacy files

Ant-hem committed:
Updated: pom.xml

Ant-hem committed:
Added: serializer test and Clojure compatibility test

Ant-hem committed:
Fixed: test suite teardown

Ant-hem committed:
Added: gzip support

Ant-hem committed:
Added: deleteby test and AATesting test

Ant-hem committed:
Added: closeable interface to clients

Ant-hem committed:
Updated: documentations

Ant-hem committed:
Added: moveIndex in SearchClient and its related test

Ant-hem committed:
Improved SearchClient with forwarding pattern

Ant-hem committed:
Added: retryStrategy tests

Ant-hem committed:
Added: requestOptions timeout

Ant-hem committed:
Code Cleanup

Ant-hem committed:
Added: sync method for SearchClientAdvanced

Ant-hem committed:
Added: searchClientAPIKeys sync methods

Ant-hem committed:
Added: searchClientCOpy sync methods

Ant-hem committed:
Added: searchClientMcm sync methods

Ant-hem committed:
Added: searchClientPersonalization sync methods

Ant-hem committed:
Added: searchIndex sync methods

Ant-hem committed:
Added: legacy properties serializer test

Ant-hem committed:
Code Cleanup + nested Arrays support for QueryString serializer

BenoitPerrot committed:
layout: Merge "main" and "test" modules

To follow the most conventional structure in the Java world.

Ant-hem committed:
Added: Java doc
- Synonyms, Rules, ABTest, Settings

Ant-hem committed:
Renamed: utils package to util

Ant-hem committed:
Renamed: klass to clazz

Ant-hem committed:
Added: static keyword to AccountClient methods

Ant-hem committed:
Added: isNullOrEmptyWhiteSpace and isEmptyWhiteSpace to AlgoliaUtils

Ant-hem committed:
Renamed: LaunderThrowable methods.
LaunderThrowable.unwrap -> LaunderThrowable.await
LaunderThrowable.launderThrowable -> LaunderThrowable.launder

Ant-hem committed:
Renamed: AlgoliaHttpRequester to HttpRequester

Ant-hem commented:
ping @BenoitPerrot ! :) Integrated some of the improvements we discussed yesterday.

Ant-hem committed:
Renamed: IHttpRequester to HttpRequester
 - AlgoliaHttpRequester to ApacheHttpRequester
 https://github.com/algolia/algoliasearch-client-java-2/issues/547

Ant-hem committed:
Renamed: AlgoliaHttpRequest to HttpRequest
- Also renamed AlgoliaHttpResponse to HttpResponse
https://github.com/algolia/algoliasearch-client-java-2/issues/546

Ant-hem committed:
Renamed: IAlgoliaWaitable response to WaitableResponse

Ant-hem committed:
Renamed: AlgoliaConfigBase to ConfigBase

Ant-hem committed:
Renamed: AlgoliaIteratorBase to IteratorBase

Ant-hem committed:
Renamed: AlgoliaIntegrationTest to IntegrationTest

Ant-hem committed:
Added: missing url parameters encoding

Ant-hem committed:
Updated: closet form signature

Ant-hem committed:
Added: Configuration Builder (#549)

* Added: Configuration Builder

* Updated: protection level of ConfigBase

* Removed: customHosts constructor in builders
- Also renamed setCustomHosts to setHosts

* Removed: customHosts and defaultHosts in favor of hosts
- Added constructor with hosts in ConfigBase
- Calling super in child classes with buildHosts() method

* Removed: unused parameters in Analytics/Insights createDefaultHosts()

* Removed: unused ApiKeys parameter in searchConfig/CreateDefault hosts

Ant-hem committed:
Added: missing parameters/Constructors
- Optional filters
- Constructor in multipleQueriesRequest and BrowseIndexQuery
- Fixed analyticsTag type

Ant-hem committed:
Added: browseSynonyms and browseRules
- Code fmt

BenoitPerrot committed:
Tests: Limit variables' scope (#550)

Ant-hem committed:
Added: addExtraHeaders to configuration builder

Ant-hem committed:
Added: personalizationImpact

BenoitPerrot committed:
tests: Split "core" and "apache" (#551)

BenoitPerrot committed:
test: Detach "core" from "apache" (#552)

* IntegrationTestExtension: Move client-independent helpers to a dedicated class

* IntegrationTestExtension: Move to "apache"

As setting up the testing environment for the Apache client.

Ant-hem committed:
Architecture: multi-modules project

PLNech commented on .run-travis.sh:
```diff
@@ -4,10 +4,10 @@ set -e
 
 if [[ "${JAVA_VERSION}" = "8" ]]; then
   if [ "$TRAVIS_PULL_REQUEST" != "false" ] && [[ ! "$TRAVIS_PULL_REQUEST_SLUG" =~ ^algolia\/ ]]; then
-    eval $(./algolia-keys export) && mvn clean test jacoco:report jacoco:report-aggregate -B && mvn -pl algoliasearch-tests coveralls:report -B;
```
Did you choose to stop reporting Code coverage?

PLNech replied:
Still `jacoco` in `pom.xml`

Ant-hem replied:
Nop I don't want to drop it! I removed it temporarily at the begining of the project. I'll add it back in a few days :)

PLNech commented on pom.xml:
```diff
@@ -1,30 +1,23 @@
-    <slf4j.version>1.7.25</slf4j.version>
-    <apache-http-client.version>4.5.5</apache-http-client.version>
-    <guava.version>24.0-jre</guava.version>
+    <async-http-client.version>4.1.4</async-http-client.version>
```
Why the regression from `4.5.5` to `4.1.4`? :thinking:

Ant-hem replied:
Because it's not the same `HttpClient` 😄 

- https://mvnrepository.com/artifact/org.apache.httpcomponents/httpclient (old one sync http client)
- https://mvnrepository.com/artifact/org.apache.httpcomponents/httpasyncclient (new one async http client)

PLNech replied:
Haha the devil is in the details :upside_down_face:

PLNech commented on src/main/java/com/algolia/search/SearchClient.java:
```diff
@@ -0,0 +1,197 @@
+    Objects.requireNonNull(indexName, "indexName can't be null.");
+
+    if (AlgoliaUtils.isEmptyWhiteSpace(indexName)) {
+      throw new NullPointerException("The index name is required");
```
On these cases I would throw an `InvalidArgumentException`, as the test does not fail only when the String is null

Ant-hem replied:
👍 I was not sure about this one. thanks

Ant-hem replied:
Fixed with 8bf447d

PLNech commented on src/main/java/com/algolia/search/util/AlgoliaUtils.java:
```diff
@@ -0,0 +1,167 @@
+   * @throws AlgoliaRuntimeException When the class doesn't have an objectID field or a Jackson
+   *     annotation @JsonProperty(\"objectID\"")
+   */
+  public static <T> void ensureObjectID(@Nonnull Class<T> clazz) {
```
Seems to duplicate the logic of `getObjectID`, which could be leveraged here :upside_down_face:

Ant-hem replied:
I'll have a look

aseure committed:
chore: change deploy step to publish to bintray rather than sonatype [WIP]

PLNech commented on src/test/java/com/algolia/search/integration/insights/InsightsTest.java:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  @Test
+  void testInsights() throws ExecutionException, InterruptedException {
```
Nothing to assert? :thinking:

Ant-hem replied:
It's how we wrote the test for the CTS. The insights API is mostly fire&forget API if we don't get a `400` from the API it's OK because we can't retrieve what was sent to check if it's the same on the server side (as we do for settings for example)

PLNech commented on src/main/java/com/algolia/search/models/HttpRequest.java:
```diff
@@ -0,0 +1,82 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class HttpRequest {
```
Could benefit Nullability annotations (same for `HttpResponse`) :)

PLNech replied:
Same for any model which would be manipulated by our users, when its fields are not covered by methods with Nullability annotations

Ant-hem replied:
Could you please elaborate? 🤔 Do you mean to add annotations on the constructors parameters?

PLNech replied:
I mean adding annotations wherever our users might benefit it:
- Constructor parameters of classes they could instantiate (not convinced it would be relevant here, as I don't think they have to create a `new HTTPRequest()` anytime)
- Annotations on public fields / public {g,s}etters that the users would manipulate to avoid unnecessary null-checks (e.g. if `headers` is never `null`, would mark `getHeaders()` as `@NonNull`) 

I think this can be applied more consistently in some classes, e.g. :
- `new Variant(...)` takes a `@NonNull String index` in constructor
-> `Variant.getIndex()` could be annotated as `@NonNull`
-> `Variant.setIndex()` should have its param annotated as `@NonNull`

You might want to use Intellij's `Infer Nullity` to quickly generate consistent annotations based on your current codebase :upside_down_face:

Ant-hem replied:
Thanks for the clarification! Neat idea. I'll look at it.

PLNech commented on src/main/java/com/algolia/search/models/apikeys/AddApiKeyResponse.java:
```diff
@@ -0,0 +1,53 @@
+            Thread.sleep(1001);
+          } catch (InterruptedException ignored) {
+          }
+          continue;
```
Isn't it a problem to ignore a `InterruptedException` and simply `continue`? :thinking: 

- https://rules.sonarsource.com/java/tag/multi-threading/RSPEC-2142
- https://stackoverflow.com/questions/28544867/is-it-ok-to-ignore-interruptedexception-if-nobody-calls-interrupt/28545020#28545020

Ant-hem replied:
👍 Good catch I'll have a look at it.

Ant-hem replied:
Fixed with 049d954 thanks for catching it

PLNech reviewed (commented):
Superficial review considering the changeset's size, but still a few (hopefully useful) comments :upside_down_face:

aseure committed:
doc: fix typos in Javadoc comments

aseure committed:
chore: remove useless Maven plugins

aseure committed:
chore: set algolia package version in a single place (in parent pom.xml)

Ant-hem committed:
Updated: POM.xml with review

- Added: classifier tests in algoliasearch-test
- Removed apache version in properties
- Removed unused assertj-core version in algoliasearch-core
- Moved findbugs in parent pom
- Moved project encoding in parent pom

aseure committed:
chore: add versions-maven-plugin to handle version changes and update release.sh script

Ant-hem committed:
Renamed: ApacheClients to DefaultClients

aseure committed:
chore: stop using versions variable in pom.xml files in favor of versions-maven-plugin set command

aseure committed:
chore: fix pom.xml files according to Maven warnings

aseure committed:
chore: ensure deploy was effectively done before git commit/tag/push in release.sh

Ant-hem committed:
Fixed: "InterruptedException" should not be ignored.

- Thread.currentThread().interrupt();
- https://rules.sonarsource.com/java/tag/multi-threading/RSPEC-2142

Ant-hem committed:
Added: missing setter for executorService in ConfigBase

Ant-hem committed:
Replaced NullPointerException with IllegalArgumentException

- When AppID or APIKey is null

Ant-hem committed:
Improved: JavaDoc for Clients

Ant-hem committed:
Added : getImplementationVersion() in ConfigBase

Ant-hem committed:
Added : ensureObjectID check in saveObjects

- Added related tests to ensureObjectID and getObjectID

Ant-hem committed:
Added: missing logType parameter to getLogs()

Ant-hem committed:
Updated: maintainers

Ant-hem committed:
Added: customRequest

PLNech reviewed (approved):
Provided you resolve the last unsolved discussions, LGTM :ok_hand:

Ant-hem commented:
> Provided you resolve the last unsolved discussions, LGTM 👌

Thanks! Still need to have a look at the Nullability annotations for some POJOs :)

Ant-hem committed:
Removed: unnecessary throws statement

Ant-hem committed:
Added: sl4j logger in HttpTransport

BenoitPerrot committed:
packaging: Generate uber jars for -core and -apache

Relocating embedded dependencies inside com.algoliasearch, filtering
out unused resources.

Note that slf4j-api is NOT embedded in the uber jars, on purpose: if
slf4j-api is embedded here, it would need to be relocated (*), yet if
gets relocated, implementations cannot be injected anymore (**),
making slf4j irrelevant.

(*) Embedding jars in an uber-library _require_ their relocation: a
library is meant to be used along with other jars, which may depend on
other versions of the jars embedded here. Relocating these mostly get
rid of conflicts; not relocating them, would on the contrary make
conflicts practically impossible to solve.

(**) the injection used by slf4j is namespace based; if the api gets
relocated, implementation would have to be relocated too.

From a "pom.xml point of view", a pluginManagement section is set in
the root pom.xml to configure the shading - but not apply it. This
configuration in then leveraged in dedicated modules
(algolia-core-uber and algolia-apache-uber) coming with their own
pom.xml files (and possibly adjusting - augmenting - the relocations,
in the apache flavor).

Ant-hem committed:
Fixed: insights test lifecycle

Ant-hem committed:
Replaced: List by Iterable in Save/Replace Synonyms/Rules

BenoitPerrot committed:
pom.xml: plugins: Remove "org.apache.maven.plugins" group id

As unnecessary, and to be consistent on all official maven plugins.

BenoitPerrot committed:
pom.xml: Put build configuration and dependencies in a profile

Activating said profile when an "src" directory can be found along the
"pom.xml" file. This way, nothing is attempted in other directories -
most notably on the root directory - avoiding useless attempts and
spurious output.

Ant-hem committed:
Renamed: browse -> browseObjects

Ant-hem committed:
Added: QueryBase with CRTP to handle Query inheritance

- Added QueryBase with CRTP to handle Query inheritance and to prevent
down-casting for the children

aseure committed:
Changed: take uber-jars POM files into account in release.sh script and ensure all steps are run in separate order

Ant-hem committed:
Renamed: Querybase -> SearchParameters

aseure committed:
chore: add maven-source-plugin to generate source jars needed by Maven Central

aseure committed:
chore: fix deploy URL to Bintray

aseure committed:
chore: generate source JARs for Uber JARS as well (mandatory for JCenter deploy)

Ant-hem committed:
Replaced SLF4J with Java.Util.logging for tracing requests/responses

Removing slf4j allow to remove on more dependency of the library and
to simplify the shading process.

 - Added: JUL in HttpTransport with Log.FINEST level
 - Removed: SLF4J and Logback from pom.xml
 - Updated: logging.properties

 To activate JUL please set in logging.properties
 .level=FINEST
 com.algolia.search.HttpTransport=FINEST
 java.util.logging.ConsoleHandler.level=FINEST

 and set:
 -Djava.util.logging.config.file="logging.properties-path"

 More information about JUL:
 http://java.ociweb.com/mark/programming/JavaLogging.html

aseure committed:
chore: upgrade maven-javadoc-plugin and maven-source-plugin to latest versions

aseure committed:
chore: force javadoc jar generation for Uber jars (needed for Maven Central publishing)

aseure committed:
chore: add missing <scm> section in parent pom.xml (required by Maven Central)

aseure committed:
chore: release of 3.0.0-beta-2 (with 1st deploy on Maven Central using Bintray!)

Ant-hem committed:
Added: missing encoding parameters in getObject()

Ant-hem committed:
Fixed: static naming conventions in configbase

Ant-hem committed:
Added: logging.properties in algoliasearch-apache

- Removed org.apache.http.level=WARNING from algoliasearch-core
- Added org.apache.http.level=WARNING in algoliasearch-apache file

Ant-hem committed:
Updated: AroundPrecision searchParameters with new format

The query parameter aroundPrecision can now handle a JSON format;
{ "aroundPrecision":
[ {"from":0, "value":10}, {"from":100, "value":1000}, ... ]
}

Ant-hem committed:
Fixed: indexing test missing future

aseure committed:
Fixed: make Consequence.setHide return a Consequence instead of void

In order to respect fluent API design used by other setters.
Reported in https://github.com/algolia/algoliasearch-client-java-2/pull/571.

aseure committed:
Fixed: make Hide.setObjectID return a Hide instead of void

In order to respect fluent API design used by other setters.
Reported in https://github.com/algolia/algoliasearch-client-java-2/pull/571.

aseure committed:
Chore: reformat files using coveo plugin

aseure committed:
Chore: remove old deploy step from the Travis script

Ant-hem committed:
Updated: Jackson version 2.9.8 > 2.9.9

Ant-hem committed:
Added: missing SearchResult properties

Fix #567

Ant-hem committed:
Removed: restrict sources in APIKey

Fix #566

Ant-hem committed:
Added: Missing settings in test settings.

Fix #569

Ant-hem committed:
Updates: README for v3

Ant-hem committed:
Adds : getConfig() in Analytics and Insights clients

Ant-hem committed:
Updated: travis.yml - JDK version
